# **Indexación y selección de datos**

En el análisis geoespacial, la capacidad de filtrar, consultar y extraer información de grandes conjuntos de datos es fundamental, ya que permite **optimizar** la manipulación de datos espaciales, **aislar** regiones de interés y mejorar la **eficiencia** en los análisis. Estas operaciones no solo facilitan la exploración de datos, sino que también **reducen la carga computacional** al trabajar con información geográfica de gran escala.

Para lograr esto, `GeoPandas` extiende las funcionalidades de `Pandas`, proporcionando métodos de **indexación y selección** que permiten **extraer subconjuntos de datos** de manera eficiente. En este artículo, exploraremos cómo realizar **selecciones** basadas en **etiquetas** y **posiciones**, así como la **selección espacial** basada en coordenadas, que facilita la extracción de datos cuyas geometrías intersectan un área delimitada. Estas herramientas son clave para estructurar consultas espaciales y optimizar el procesamiento de datos geográficos.

Comenzaremos importando las librerías necesarias y cargando los datos que utilizaremos en este tutorial.

In [1]:
# importar librerías
import numpy as np
import geopandas as gpd
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Ruta del archivo
uri=r'D:\Charlie\01_Cartografia\catastro.gpkg'

# Lectura como GeoDataFrame
puertas = gpd.read_file(uri, layer='puertas')

In [3]:
# Visualizar las dos primera filas
puertas.head(2)

OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
0        61          L38       None          842         150132   
1        62          L24       None         811A         150132   

  CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
0            147461        31856.0          3687   
1            147899        31856.0          3688   

                         geometry  
0  POINT (282306.453 8669914.377)  
1  POINT (282245.352 8669863.459)

Veamos las columnas de esta capa:

In [4]:
puertas.columns.values

array(['OBJECTID', 'CODIGOPREDIO', 'NUMEROLOTE', 'NUMEROPUERTA',
       'CODIGODISTRITO', 'CODIGOSEGMENTOVIA', 'CODIGOMANZANA',
       'CODIGOPUERTA', 'geometry'], dtype=object)

## **1. Selección con Operadores de Indexación**

Para seleccionar datos, podemos usar los operadores `[]` y `.`, que permiten **acceder a columnas** específicas y **filtrar subconjuntos** de datos mediante **máscaras booleanas**.

### **1.1. Selección de columnas**

#### **Selección de una columna específica**

Para acceder a una columna específica, debemos indicar su **etiqueta** de la siguiente manera:

```{tip} Seleccionar una columna específica
df[column1]
```

Por ejemplo, para seleccionar la columna `NUMEROPUERTA`:

In [5]:
# Seleccionar la columna NUMEROPUERTA
puertas['NUMEROPUERTA']

0           842
1          811A
2           930
3           943
4           971
           ... 
2323691     S/N
2323692     S/N
2323693     S/N
2323694     S/N
2323695     S/N
Name: NUMEROPUERTA, Length: 2323696, dtype: object

También podemos usar el operador `.`, siempre que el nombre de la columna no contenga espacios ni caracteres especiales.

```python
df.column1
```

De esta forma, también podemos seleccionar la columna `NUMEROPUERTA`.

In [6]:
# Seleccionar la columna NUMEROPUERTA usando "."
puertas.NUMEROPUERTA

0           842
1          811A
2           930
3           943
4           971
           ... 
2323691     S/N
2323692     S/N
2323693     S/N
2323694     S/N
2323695     S/N
Name: NUMEROPUERTA, Length: 2323696, dtype: object

#### **Selección de multiples columnas**

Para seleccionar múltiples columnas, debemos proporcionar una **lista con sus etiquetas**:

```python
df[[column1, column2, ..., columnN]]
```

Por ejemplo, así seleccionamos las columnas `NUMEROPUERTA`, `CODIGODISTRITO` y `geometry`:

In [7]:
# Selección de columnas en base a una lista de etiquetas
puertas[['NUMEROPUERTA', 'CODIGODISTRITO', 'geometry']]

NUMEROPUERTA CODIGODISTRITO                        geometry
0                842         150132  POINT (282306.453 8669914.377)
1               811A         150132  POINT (282245.352 8669863.459)
2                930         150132  POINT (282501.278 8670013.832)
3                943         150132  POINT (282505.574 8669977.579)
4                971         150132   POINT (282560.12 8669989.886)
...              ...            ...                             ...
2323691          S/N         150112  POINT (278698.773 8676598.118)
2323692          S/N         150112  POINT (278682.456 8676625.591)
2323693          S/N         150112  POINT (278677.124 8676632.768)
2323694          S/N         150112  POINT (278670.622 8676643.207)
2323695          S/N         150112  POINT (278666.108 8676648.642)

[2323696 rows x 3 columns]

### **1.2. Selección con máscaras boleanas**

La selección con máscaras booleanas permite filtrar datos mediante **expresiones lógicas**. Se genera una **Serie boleana** donde **`True`** indica las **filas a conservar** y **`False`** las que **se descartan**.

```python
df[<expresiones_lógicas>]
```

Por ejemplo, para seleccionar las filas donde la edad sea mayor a 30, usamos `df[df['edad'] > 30]`, donde la expresión `df['edad'] > 30` genera una máscara booleana que mantiene únicamente las filas que cumplen esta condición.

Ahora, seleccionemos las puertas cuyo código de distrito sea igual a "150117":

In [8]:
# Máscara donde el codigo de distrito es "150117"
mascara = puertas['CODIGODISTRITO']=='150117'

# Extración de los registros donde CODIGODISTRITO es '150117'
puertas_olivos = puertas[mascara]

# Imprimir la cantidad de filas y columnas extraidas
print(f'Fila y columnas: {puertas_olivos.shape}')

# Visualizar las dos primeras filas:
puertas_olivos.head(2)

Fila y columnas: (54401, 9)


OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
408797    594697          L18          0          S/N         150117   
408798    594698          L01          6          S/N         150117   

       CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
408797            239440        44601.0        837603   
408798            239995        44593.0        837604   

                              geometry  
408797  POINT (273398.254 8671700.521)  
408798  POINT (273837.064 8671477.364)

Podemos combinar varias **expresiones lógicas** utilizando `&` (**AND**), `|` (**OR**) y `~` (**NOT**). Es importante usar paréntesis para definir la prioridad de evaluación. También es recomendable almacenar cada condición en una variable:

In [9]:
# Mask 1: Registros cuyo CODIGODISTRITO es '150117'
maskDist = puertas.CODIGODISTRITO=='150117'

# Mask 2: Registros cuyo NUMEROPUERTA es diferente a "S/N"
maskPuerta = puertas.NUMEROPUERTA != 'S/N'

# Seleccionando por las condiciones:
puertas_olivos_names = puertas[maskDist & maskPuerta]

# Imprimir la cantidad de filas y columnas seleccionadas
print(f'Fila y columnas: {puertas_olivos_names.shape}')

# Visualizar las dos primeras filas:
puertas_olivos_names.head(2)

Fila y columnas: (19965, 9)


OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
408847    594747          L13       None          319         150117   
408857    594757          L08       None         1888         150117   

       CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
408847            239305        44595.0        837653   
408857            239193        44602.0        837663   

                              geometry  
408847  POINT (273774.795 8671638.546)  
408857   POINT (273334.322 8671831.08)

### **1.3. Integrando selección de columnas y por máscara boleana**

Mediante operadores de indexación, podemos combinar la **selección de columnas** y el **filtrado con máscaras booleanas** en una sola operación.

In [10]:
# Lista de columnas a seleccionar
cols = ['NUMEROPUERTA', 'CODIGODISTRITO', 'geometry']

# Filtramos usando las variables "maskDist", "maskPuerta" y "cols"
puertas_select = puertas[maskDist & maskPuerta][cols]

# Imprimir la cantidad de filas y columnas filtradas
print(f'Fila y columnas: {puertas_select.shape}')

# Visualizar las dos primeras filas:
puertas_select.head(2)

Fila y columnas: (19965, 3)


NUMEROPUERTA CODIGODISTRITO                        geometry
408847          319         150117  POINT (273774.795 8671638.546)
408857         1888         150117   POINT (273334.322 8671831.08)

## **2. Selección por etiqueta `.loc`**

La selección por etiqueta se realiza mediante el método _Pandas_ **[.loc](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html)**, que permite **acceder a filas y columnas** a través de sus **nombres de índice**.

### **2.1. Selección de filas**

Las filas se seleccionan utilizando **etiquetas de filas**, que por defecto son números enteros, a menos que se haya definido otro valor. Si no se especifica ninguna columna, se devolverán todas por defecto.

#### **Selección de una Fila Específica**

Para seleccionar una única fila, utilizamos:

```python
df.loc[row1]
```

Esto devuelve todos los valores de la fila con la etiqueta `row1`.

In [11]:
# Seleccionando la etiqueta de fila 5:
puertas.loc[5]

OBJECTID                                                 66
CODIGOPREDIO                                            L17
NUMEROLOTE                                             None
NUMEROPUERTA                                            528
CODIGODISTRITO                                       150132
CODIGOSEGMENTOVIA                                    159355
CODIGOMANZANA                                       31811.0
CODIGOPUERTA                                           3692
geometry             POINT (282541.47200000007 8670064.495)
Name: 5, dtype: object

#### **Selección de múltiples filas**

Para seleccionar múltiples filas, utilizamos:

```python
df.loc[[row1, row2]]
```

Retorna un **subconjunto de datos** que incluye las filas con las etiquetas `row1` y `row2`.

In [12]:
# Seleccionando un lista de etiquetas de filas
puertas.loc[[5,10,37]]

OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
5         66          L17       None          528         150132   
10        71          L15       None         819A         150132   
37        98          L10       None        1029B         150132   

   CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
5             159355        31811.0          3692   
10            152388        31852.0          3697   
37            156930        31818.0          3724   

                          geometry  
5   POINT (282541.472 8670064.495)  
10  POINT (282279.506 8669810.186)  
37   POINT (282732.879 8669959.43)

#### **Selección mediante `slicing`**

También podemos utilizar un **rango de etiquetas** de fila con `slicing`:

```python
df.loc[rowA:rowN]
```

Esto retorna un **subconjunto de datos** que incluye todas las filas desde la etiqueta `rowA` hasta `rowN`, incluyéndolas.

In [13]:
# Seleccionando etiquetas de filas con un slice 
puertas.loc[40:42]

OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
40       101          L10       None          547         150132   
41       102          L17       None          100         150132   
42       103          L47       None          S/N         150132   

   CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
40            162589        31625.0          3727   
41            155519        31625.0          3728   
42                -1        31617.0          3729   

                          geometry  
40  POINT (281663.218 8670001.857)  
41  POINT (281605.177 8669902.147)  
42  POINT (281436.459 8669876.181)

### **2.2. Selección de columnas**

Las columnas se seleccionan mediante su **etiqueta de índice**, que corresponde a su **nombre**. Sin embargo, también es **necesario especificar las filas**. Para seleccionar **todas las filas**, podemos utilizar el operador `:`.

#### **Selección de una columna específica**

Para seleccionar una única columna, utilizamos:

```python
df.loc[:, column1]
```

Es necesario especificar las filas, pero podemos usar el operador `:` para seleccionar todas. Esto devuelve la columna `column1` con todos sus valores.

In [14]:
# Selección de la columna "NUMEROPUERTA" y todas sus filas
puertas.loc[:,'NUMEROPUERTA']

0           842
1          811A
2           930
3           943
4           971
           ... 
2323691     S/N
2323692     S/N
2323693     S/N
2323694     S/N
2323695     S/N
Name: NUMEROPUERTA, Length: 2323696, dtype: object

#### **Selección de múltiples columnas**

Para seleccionar múltiples columnas, pasamos una **lista con sus etiquetas** de la siguiente manera:

```python
df.loc[:, [column1, column2]]
```

Esto devuelve un subconjunto de datos con las columnas `column1` y `column2`.

In [15]:
# Selección de múltiples columnas y todas su filas
puertas.loc[:, ['NUMEROPUERTA','CODIGODISTRITO','geometry']]

NUMEROPUERTA CODIGODISTRITO                        geometry
0                842         150132  POINT (282306.453 8669914.377)
1               811A         150132  POINT (282245.352 8669863.459)
2                930         150132  POINT (282501.278 8670013.832)
3                943         150132  POINT (282505.574 8669977.579)
4                971         150132   POINT (282560.12 8669989.886)
...              ...            ...                             ...
2323691          S/N         150112  POINT (278698.773 8676598.118)
2323692          S/N         150112  POINT (278682.456 8676625.591)
2323693          S/N         150112  POINT (278677.124 8676632.768)
2323694          S/N         150112  POINT (278670.622 8676643.207)
2323695          S/N         150112  POINT (278666.108 8676648.642)

[2323696 rows x 3 columns]

#### **Selección mediante `slicing`**

También podemos emplear un **rango de etiquetas** para seleccionar columnas mediante `slicing`:

```python
df.loc[columnA:columnN]
```

Esto devuelve un **subconjunto de datos** que incluye todas las columnas desde `columnA` hasta `columnN`.

In [16]:
puertas.loc[:, 'CODIGOPREDIO':'NUMEROPUERTA']

CODIGOPREDIO NUMEROLOTE NUMEROPUERTA
0                L38       None          842
1                L24       None         811A
2                L21       None          930
3                L13       None          943
4                L06       None          971
...              ...        ...          ...
2323691         L010          1          S/N
2323692         L021          4          S/N
2323693         L040          5          S/N
2323694         L030          6          S/N
2323695         L020          7          S/N

[2323696 rows x 3 columns]

### **2.3. Selección de filas y columnas**

Una de las principales ventajas de **`loc`** es su capacidad para **seleccionar filas y columnas** simultáneamente de manera intuitiva, utilizando **etiquetas en ambas dimensiones**.

#### **Selección de una fila y una columna**

Para seleccionar un único valor, especificamos la **etiqueta** de la **fila** y la de la **columna**:

```python
df.loc[row1, column1]
```

Esto devuelve el valor específico ubicado en la **intersección** de la fila `row1` y la columna `column1`.

In [17]:
# Seleccionando el valor almacenado en la fila 5
# de la columna "NUMEROPUERTA"
puertas.loc[5, 'NUMEROPUERTA']

'528'

#### **Selección de mútiples filas y columnas**

Para seleccionar múltiples filas y columnas, debemos proporcionar una **lista con sus etiquetas de índice** en ambos ejes:

```python
df.loc[[row1, row2], [column1, column2]]
```

Esto devuelve un subconjunto de datos que incluye únicamente las **filas** `row1` y `row2`, junto con las **columnas** `column1` y `column2`.

In [18]:
puertas.loc[[5,6,7,8], ['CODIGOPREDIO','NUMEROPUERTA']]

CODIGOPREDIO NUMEROPUERTA
5          L17          528
6          L11          529
7          L10          535
8          L01          273

#### **Selección mediante `slicing`**

También podemos emplear un **rango de etiquetas** para seleccionar **filas** y **columnas** mediante `slicing`:

```python
df.loc[rowA:rowN, columnA:columnN]
```

Esto devuelve un **subconjunto de datos** que abarca las **filas** desde `rowA` hasta `rowN` y las **columnas** desde `columnA` hasta `columnN`, incluyéndolas.

In [19]:
puertas.loc[5:8, 'CODIGOPREDIO':'NUMEROPUERTA']

CODIGOPREDIO NUMEROLOTE NUMEROPUERTA
5          L17       None          528
6          L11       None          529
7          L10       None          535
8          L01       None          273

### **2.4. Selección con máscaras boleanas**

El método  **`loc`** permite seleccionar datos utilizando máscaras booleanas, generadas a partir de **expresiones lógicas**. Estas devuelven una **Serie boleana** en la que `True` indica las **filas a conservar** y `False`, las **filas a descartar**.

```python
df.loc[<expresiones_lógicas>]
```

A continuación, seleccionaremos todas las puertas sin número asignado y mostraremos únicamente las columnas `CODIGOPREDIO` y `NUMEROPUERTA`.

In [20]:
# Mascara de NUMEROPUERTA igual a sin nombre
maskPuertas = puertas.NUMEROPUERTA == 'S/N'

# Lista de columnas a seleccionar
colsPuertas = ['CODIGOPREDIO','NUMEROPUERTA']

# Filtro por condición y de columnas
puertas.loc[maskPuertas, colsPuertas]

CODIGOPREDIO NUMEROPUERTA
9                L16          S/N
12               L44          S/N
14               L19          S/N
16               L40          S/N
17               L23          S/N
...              ...          ...
2323691         L010          S/N
2323692         L021          S/N
2323693         L040          S/N
2323694         L030          S/N
2323695         L020          S/N

[1518892 rows x 2 columns]

Podemos generar estadísticas descriptivas del campo `NUMEROPUERTA`:

In [21]:
# Cálculo de estadísticas descriptivas
puertas.NUMEROPUERTA.describe()

count     2323696
unique      21137
top           S/N
freq      1518892
Name: NUMEROPUERTA, dtype: object

Las estadísticas del campo `NUMEROPUERTA` indican que hay 2,323,696 valores no nulos, pero este conteo incluye el valor **"S/N"**, que es el más frecuente, con 1,518,892 apariciones. Sin embargo, **"S/N"** no representa un número de puerta válido, por lo que debería reemplazarse por `NaN` para obtener un análisis más preciso.

En el siguiente apartado, aplicaremos esta limpieza asignando valores a una columna específica.

### **2.5. Asignación de nuevo valores**

Una de las principales ventajas de loc es su capacidad para asignar valores a una o varias columnas, siempre que las filas cumplan una determinada condición, es decir, aquellas donde las **expresiones lógicas** evalúe como `True`.

```python
df.loc[<expresiones_lógicas>, <columnas_modificar>] = <nuevo_valor>
```

En el apartado anterior, identificamos las filas donde `NUMEROPUERTA` no tiene un valor válido, lo que puede afectar los análisis de completitud. En este ejemplo, reemplazaremos esos valores por `NaN` en la columna `NUMEROPUERTA` para mejorar la calidad de los datos.

```{note}
Es importante especificar correctamente las columnas, ya que los cambios se aplicarán a todas las columnas que indiquemos. Si no se especifica la columna, la modificación afectará a todas las columnas del **GeoDataFrame**.
```

In [22]:
# Reemplazando los NUMEROPUERTA sin nombre por NaN
puertas.loc[puertas.NUMEROPUERTA == 'S/N', 'NUMEROPUERTA'] = np.nan

Volvamos a cacular las estadísticas descriptivas para el campo `NUMEROPUERTA`:

In [23]:
# Cálculo de estadísticas descriptivas
puertas.NUMEROPUERTA.describe()

count     804804
unique     21136
top          120
freq        3108
Name: NUMEROPUERTA, dtype: object

Ahora, extraeremos el subconjunto de datos que si tiene `NUMEROPUERTA`

In [24]:
puertas_validas = puertas[puertas.NUMEROPUERTA.notna()]
puertas_validas.shape

(804804, 9)

La asignación de valores también nos permite **crear nuevas columnas** para almacenar información adicional. Por ejemplo, podemos agregar la columna `PARIDAD` para indicar si el número de puerta es par o impar. Aplicaremos esta asignación sobre el nuevo `GeoDataFrame` **puertas_validas**.

In [25]:
# Condicion para evaluar puertas pares
conPares = puertas_validas.NUMEROPUERTA.str.extract(r"(\d+)")[0].str[-1]\
                          .isin([str(i) for i in list(range(0,9,2))])

# Condicion para evaluar puertas impares
conNones = puertas_validas.NUMEROPUERTA.str.extract(r"(\d+)")[0].str[-1]\
                          .isin([str(i) for i in list(range(1,10,2))])

# Aplicando las condiciones para calcular un nuevo campo "Paridad"
puertas_validas.loc[conPares,'PARIDAD'] = 'Pares'
puertas_validas.loc[conNones,'PARIDAD'] = 'Impares'

# Visualizando:
puertas_validas[['CODIGOPREDIO','NUMEROPUERTA','PARIDAD']]

CODIGOPREDIO NUMEROPUERTA  PARIDAD
0                L38          842    Pares
1                L24         811A  Impares
2                L21          930    Pares
3                L13          943  Impares
4                L06          971  Impares
...              ...          ...      ...
2323626         L030          254    Pares
2323629         L08A          117  Impares
2323630          L22         214B    Pares
2323632         L071          170    Pares
2323633         L060          146    Pares

[804804 rows x 3 columns]

## **3. Selección por posición `.iloc`**

La selección por posición se realiza mediante el método _Pandas_ **[.iloc](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html)**, que permite **acceder a filas y columnas** mediante **índices enteros basados en su posición**.

### **3.1. Selección de filas**

Las filas se seleccionan utilizando su **posición numérica**. Si no se especifica ninguna columna, se devolverán todas por defecto.


#### **Selección de una Fila Específica**

Para seleccionar una única fila, utilizamos:

```python
df.iloc[i]
```

Esto devuelve todos los valores de la fila en la posición **`i`**.

In [26]:
# Retorna la posición 0
puertas.iloc[0]

OBJECTID                                                 61
CODIGOPREDIO                                            L38
NUMEROLOTE                                             None
NUMEROPUERTA                                            842
CODIGODISTRITO                                       150132
CODIGOSEGMENTOVIA                                    147461
CODIGOMANZANA                                       31856.0
CODIGOPUERTA                                           3687
geometry             POINT (282306.45299999975 8669914.377)
Name: 0, dtype: object

#### **Selección de múltiples filas**

Para seleccionar múltiples filas, utilizamos:

```python
df.iloc[[i, j, k]]
```

Esto devuelve un **subconjunto de datos** con las filas en las posiciones `i`, `j` y `k`.

In [27]:
# Retorna las fila en la posición 0, 10 y 15
puertas.iloc[[0, 10, 15]]

OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
0         61          L38       None          842         150132   
10        71          L15       None         819A         150132   
15        76          L20       None          180         150132   

   CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
0             147461        31856.0          3687   
10            152388        31852.0          3697   
15            154840        31849.0          3702   

                          geometry  
0   POINT (282306.453 8669914.377)  
10  POINT (282279.506 8669810.186)  
15  POINT (282470.515 8669747.199)

#### **Selección mediante `slicing`**

También podemos utilizar un **rango de indices** para seleccionar filas mediante `slicing`:

```python
df.iloc[i:k]
```

Esto devuelve un **subconjunto de datos** incluyendo las filas desde la posición `i` hasta `k-1`.

In [28]:
# Retorna las filas desde la posicion 5 al 8
puertas.iloc[5:9]

OBJECTID CODIGOPREDIO NUMEROLOTE NUMEROPUERTA CODIGODISTRITO  \
5        66          L17       None          528         150132   
6        67          L11       None          529         150132   
7        68          L10       None          535         150132   
8        69          L01       None          273         150132   

  CODIGOSEGMENTOVIA  CODIGOMANZANA  CODIGOPUERTA  \
5            159355        31811.0          3692   
6            150068        31811.0          3693   
7            150068        31811.0          3694   
8            157623        31854.0          3695   

                         geometry  
5  POINT (282541.472 8670064.495)  
6  POINT (282575.181 8670074.712)  
7  POINT (282572.823 8670082.021)  
8  POINT (282423.179 8669852.588)

### **3.2. Selección de columnas**

Las columnas se seleccionan mediante su **posición numérica**, es decir, su índice entero. Asimismo, es **necesario especificar las filas**. Para seleccionar **todas las filas**, utilizamos el operador `:`.

#### **Selección de una columna específica**

Para seleccionar una única columna por su posición, usamos:

```python
df.iloc[:, i]
```

Esto devuelve la columna ubicada en la posición `i`.

In [29]:
# Retorna la columna en la posición 3 (NUMEROPUERTA)
puertas.iloc[:, 3]

0           842
1          811A
2           930
3           943
4           971
           ... 
2323691     NaN
2323692     NaN
2323693     NaN
2323694     NaN
2323695     NaN
Name: NUMEROPUERTA, Length: 2323696, dtype: object

#### **Selección de múltiples columnas**

Para seleccionar múltiples columnas por su posición, pasamos una **lista con sus índices** de la siguiente manera:

```python
df.iloc[:, [i, j]]
```

Esto devuelve un **subconjunto de datos** con las columnas ubicadas en las posiciones `i` y `j`.

In [30]:
# Retorna la columna en las posiciones 0, 3, -1  
# (CODIGOPREDIO, NUMEROPUERTA, geometry)
puertas.iloc[:, [0, 3, -1]]

OBJECTID NUMEROPUERTA                        geometry
0              61          842  POINT (282306.453 8669914.377)
1              62         811A  POINT (282245.352 8669863.459)
2              63          930  POINT (282501.278 8670013.832)
3              64          943  POINT (282505.574 8669977.579)
4              65          971   POINT (282560.12 8669989.886)
...           ...          ...                             ...
2323691   3717000          NaN  POINT (278698.773 8676598.118)
2323692   3717001          NaN  POINT (278682.456 8676625.591)
2323693   3717002          NaN  POINT (278677.124 8676632.768)
2323694   3717003          NaN  POINT (278670.622 8676643.207)
2323695   3717004          NaN  POINT (278666.108 8676648.642)

[2323696 rows x 3 columns]

#### **Selección mediante `slicing`**

También podemos emplear un **rango de indices** para seleccionar columnas mediante `slicing`:

```python
df.iloc[:, i:k]
```

Esto devuelve un **subconjunto de datos** que incluye todas las columnas desde la fila `i` hasta `k-1`.

In [32]:
# Retorna la columna en la posicion 1, 2 y 3
puertas.iloc[:, 1:4]

CODIGOPREDIO NUMEROLOTE NUMEROPUERTA
0                L38       None          842
1                L24       None         811A
2                L21       None          930
3                L13       None          943
4                L06       None          971
...              ...        ...          ...
2323691         L010          1          NaN
2323692         L021          4          NaN
2323693         L040          5          NaN
2323694         L030          6          NaN
2323695         L020          7          NaN

[2323696 rows x 3 columns]

### **3.3. Selección de filas y columnas**

### **3.4. Selección con máscaras boleanas**

### **3.5. Asignación de nuevos valores**

## **4. Selección por cuadro delimitador `.cx`**